generate-stackbrew-library.sh

```diff
@@ -62,57 +29,73 @@ Maintainers: Tianon Gravi <admwiggin@gmail.com> (@tianon),
              Joseph Ferguson <yosifkit@gmail.com> (@yosifkit),
              Austin Burdine <austin@acburdine.me> (@acburdine)
 GitRepo: https://github.com/docker-library/ghost.git
+GitCommit: $commit
 EOH
 
-# prints "$2$1$3$1...$N"
-join() {
-	local sep="$1"; shift
-	local out; printf -v out "${sep//%/%%}%s" "$@"
-	echo "${out#$sep}"
-}
-
-for version in "${versions[@]}"; do
-	rcVersion="${version%-rc}"
-
-	for variant in debian alpine; do
-		commit="$(dirCommit "$version/$variant")"
-
-		fullVersion="$(git show "$commit":"$version/$variant/Dockerfile" | awk '$1 == "ENV" && $2 == "GHOST_VERSION" { print $3; exit }')"
-
-		versionAliases=()
-		if [ "$version" = "$rcVersion" ]; then
-			while [ "$fullVersion" != "$version" -a "${fullVersion%[.-]*}" != "$fullVersion" ]; do
-				versionAliases+=( $fullVersion )
-				fullVersion="${fullVersion%[.-]*}"
-			done
-		fi
-		versionAliases+=(
-			$fullVersion
-			${aliases[$version]:-}
-		)
-
-		if [ "$variant" = "$defaultVariant" ]; then
-			variantAliases=( "${versionAliases[@]}" )
-		else
-			variantAliases=( "${versionAliases[@]/%/-$variant}" )
-			variantAliases=( "${variantAliases[@]//latest-/}" )
-		fi
-
-		variantParent="$(awk 'toupper($1) == "FROM" { print $2 }' "$version/$variant/Dockerfile")"
-		variantArches="${parentRepoToArches[$variantParent]}"
-
-		if [ "$variant" = 'alpine' ]; then
-			# ERROR: unsatisfiable constraints:
-			#   vips-dev (missing):
-			variantArches="$(sed -e 's/ ppc64le / /g' -e 's/ s390x / /g' <<<" $variantArches ")"
-		fi
-
-		echo
-		cat <<-EOE
-			Tags: $(join ', ' "${variantAliases[@]}")
-			Architectures: $(join ', ' $variantArches)
-			GitCommit: $commit
-			Directory: $version/$variant
-		EOE
-	done
-done
+jq --raw-output --argjson parentArches "$parentArchesJson" '
+	# explodes a version to a list of tags
+	# usage: .version | tags
+	# e.g. "1.2.3.4" | tags -> [ "1.2.3.4", "1.2.3", "1.2", "1" ]
+	def tags:
+		# TODO improve rc suffix support (e.g. this wont work for -beta or -rc.1 versioned software)
+		(if endswith("-rc") then "-rc" else "" end) as $suffix
+		| rtrimstr($suffix)
+		| split(".")
+		| [ foreach .[] as $c ([]; .+=[$c]) | join(".") + $suffix ]
+		| reverse
+	;
+
+	# returns a list of tags for this variant
+	# usage: variant_tags("version"; [sortedVersionList]; "variant"; [sortedVariantList])
+	# e.g. variant_tags("1.2.3.4"; ["1.2.3.4", "0.1.2.3"]; "trixie"; ["trixie", "alpine3.22"]) -> ["1.2.3.4-trixie", "1.2.3-trixie", ...]
+	def variant_tags($version; $versionListSorted; $variant; $variantListSorted):
```

Comment:
Given that each of these are only used once, what's the benefit to having functions over just embedding the logic below? :sweat_smile: